pkg/controller/stacks/persona/persona.go

```diff
@@ -0,0 +1,260 @@
+	for _, persona := range personas {
+		name := fmt.Sprintf(stacks.NamespaceClusterRoleNameFmt, nsName, persona)
+
+		labels := map[string]string{
+			fmt.Sprintf(stacks.LabelNamespaceFmt, nsName): "true",
+			stacks.LabelScope: stacks.NamespaceScoped,
+		}
+
+		if persona == adminPersona {
+			labels[fmt.Sprintf(stacks.LabelAggregateFmt, "crossplane", persona)] = "true"
+		}
+
+		// By specifying MatchLabels, ClusterRole Aggregation will pass
+		// along the rules from other ClusterRoles with the desired labels.
+		// This is why we don't define any Rules here.
+		role := &rbacv1.ClusterRole{
+			AggregationRule: &rbacv1.AggregationRule{
+				ClusterRoleSelectors: []metav1.LabelSelector{
+					{
+						MatchLabels: map[string]string{
+							fmt.Sprintf(stacks.LabelAggregateFmt, stacks.NamespaceScoped, persona): "true",
+							fmt.Sprintf(stacks.LabelNamespaceFmt, nsName):                          "true",
+						},
+					},
+					{
+						MatchLabels: map[string]string{
+							fmt.Sprintf(stacks.LabelAggregateFmt, "namespace-default", persona): "true",
+						},
+					},
+				},
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   name,
+				Labels: labels,
+			},
+		}
+
+		// Parent labels enable more precise DeleteAllOf calls than matching
+		// on clusterrole name
+		meta.AddLabels(role, stacks.ParentLabels(ns))
+
+		roles = append(roles, role)
+	}
+
+	return roles
+}
+
+func nsHasPersonaManagement(ns *corev1.Namespace) bool {
+	v, ok := ns.GetLabels()[managedRolesLabel]
+	return ok && v == managedRolesEnabled
+}
+
+// create ClusterRoles for namespace personas
+// example: crossplane:ns:{name}:{persona}
+func (h *nsPersonaHandler) create(ctx context.Context) error {
+	roles := generateNamespaceClusterRoles(h.ns)
+
+	for _, role := range roles {
+		// When the namespace is deleted, clusterroles are no longer needed.
+		// Set the owner to the Namespace for garbage collection.
+		role.SetOwnerReferences([]metav1.OwnerReference{
+			meta.AsOwner(meta.ReferenceTo(h.ns, corev1.SchemeGroupVersion.WithKind("Namespace"))),
+		})
+
+		// Creating the clusterroles. Rules in these clusterroles are populated
+		// through aggregation from the stacks installed in the namespaces, we
+		// won't need to update them.
+		//
+		// We are not patching existing clusterroles. This permits user
+		// modification.
+		h.log.Debug("Creating ClusterRole", "name", role.GetName())
+		if err := h.kube.Create(ctx, role); err != nil && !kerrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, errFailedToCreateClusterRole)
+		}
+	}
+	return nil
+}
+
+// delete ClusterRoles for namespace personas
+func (h *nsPersonaHandler) delete(ctx context.Context) error {
+	// Logging that clusterroles are attempting to be deleted would
+	// either be noisy (logged on unmanaged namespaces) or cost a
+	// lookup for existing clusterroles
+
+	labels := stacks.ParentLabels(h.ns)
+	if err := h.kube.DeleteAllOf(ctx, &rbacv1.ClusterRole{}, client.MatchingLabels(labels)); err != nil {
```

Comment:
I don't think DeleteAllOf returns `NotFound` if already deleted, but worths checking. We would need to ignore `NotFound` in that case.

Reply:
I've verified that if I delete the clusterroles before removing the label there is no error condition here.

I believe I was wary of this same possible error value in a past PR and may have included a `NotFound` check.  I'll remove that when I next encounter that code.